optuna/_transform.py

```diff
@@ -0,0 +1,271 @@
+            else:
+                param = _untransform_numerical_param(
+                    trans_param.item(), distribution, self._transform_log
+                )
+
+            params[name] = param
+
+        return params
+
+
+def _transform_search_space(
+    search_space: Dict[str, BaseDistribution], transform_log: bool, transform_step: bool
+) -> Tuple[numpy.ndarray, List[numpy.ndarray]]:
+    assert len(search_space) > 0, "Cannot transform if no distributions are given."
+
+    n_bounds = sum(
+        len(d.choices) if isinstance(d, CategoricalDistribution) else 1
+        for d in search_space.values()
+    )
+
+    bounds = numpy.empty((n_bounds, 2), dtype=numpy.float64)
+    column_to_encoded_columns: List[numpy.ndarray] = []
+
+    bound_idx = 0
+    for distribution in search_space.values():
+        d = distribution
+        if isinstance(d, CategoricalDistribution):
+            n_choices = len(d.choices)
+            bounds[bound_idx : bound_idx + n_choices] = (0, 1)  # Broadcast across all choices.
+            column_to_encoded_columns.append(numpy.arange(bound_idx, bound_idx + n_choices))
+            bound_idx += n_choices
+        elif isinstance(
+            d,
+            (
+                UniformDistribution,
+                LogUniformDistribution,
+                DiscreteUniformDistribution,
+                IntUniformDistribution,
+                IntLogUniformDistribution,
+            ),
+        ):
+            if isinstance(d, UniformDistribution):
+                bds = (
+                    _transform_numerical_param(d.low, d, transform_log),
+                    _transform_numerical_param(d.high, d, transform_log),
+                )
+            elif isinstance(d, LogUniformDistribution):
+                bds = (
+                    _transform_numerical_param(d.low, d, transform_log),
+                    _transform_numerical_param(d.high, d, transform_log),
+                )
+            elif isinstance(d, DiscreteUniformDistribution):
+                half_step = 0.5 * d.q if transform_step else 0.0
+                bds = (
+                    _transform_numerical_param(d.low, d, transform_log) - half_step,
+                    _transform_numerical_param(d.high, d, transform_log) + half_step,
+                )
+            elif isinstance(d, IntUniformDistribution):
+                half_step = 0.5 * d.step if transform_step else 0.0
+                bds = (
+                    _transform_numerical_param(d.low, d, transform_log) - half_step,
+                    _transform_numerical_param(d.high, d, transform_log) + half_step,
+                )
+            elif isinstance(d, IntLogUniformDistribution):
+                half_step = 0.5 if transform_step else 0.0
+                bds = (
+                    _transform_numerical_param(d.low - half_step, d, transform_log),
+                    _transform_numerical_param(d.high + half_step, d, transform_log),
+                )
```

Comment:
I think it's difficult given that the steps are applied differently depending on the distribution, more precisely `IntLogUniformDistribution`. Could you elaborate on how?

Reply:
I thought the following implementation of `_transform_numerical_param`. I'm suggesting this because I think it would be easier to read the code if the case for distribution is localized. However, this is my preference, so I think there is a room of discussion.
```python
def _transform_numerical_param(
    param: Union[int, float], 
    distribution: BaseDistribution, 
    transform_log: bool, 
    transform_step: bool, 
    position: str
) -> float:
    d = distribution
    assert position is in ("low", "high", "inside")
    
    ...
    if isinstance(d, DiscreteUniformDistribution):
        if position == "low"
            extended_step = -0.5 * d.q if transform_step else 0.0
        elif position == "high"
            extended_step = 0.5 * d.q if transform_step else 0.0
        else:
            extended__step = 0
        
        trans_param = float(param) + extended_step
    elif ....
```

Reply:
Oh, thank you for the detailed code. I now see what you mean, passing the position as well. It might make `_transform_search_space` more readable but I'm not sure it contributes to the overall readability. `_transform_numerical_param` is obviously more complicated and moreover it's no longer symmetrical to `_untransform_numerical_param`.

Reply:
OK, I see. Let me resolve this conversation without any fixes.